feat: use typescript to parse js blocks for `Hover`

ashish10alex opened the pull request:
This PR:

Updates https://github.com/ashish10alex/vscode-dataform-tools/pull/58 instead to use [typescripts](https://www.npmjs.com/package/typescript) ability to parse javascript code and give metadata about each node. typescript is already a dev dependancy in our repo I have added this as a dependancy for the user as well as we can use this for hover and to refactor go to definition in the future. 



Please see the working example video. I used the test code you provided but had to modify some columns as they were giving Dataform compilation error. 



https://github.com/user-attachments/assets/78bd13c7-ede8-4c43-a12c-95445d11a196

HampB commented:
Thank you for changes @ashish10alex, this makes it alot cleaner and more stable. For sure this is the way to go.

An issue though, as can be seen in your screen capture aswell, is that we do not infer the types for a function any more. 
We also do not get the param names in the doc-string, we just get the tagName and description. 
![image](https://github.com/user-attachments/assets/c890ae1f-a2c6-4e78-911d-ffd016cd4971)

A quick google search tells me that that might not be possible with the methods used in this PR. 
the node.parameters[*].type will return undefined as long as the type is not specified like this:
![image](https://github.com/user-attachments/assets/ae550b30-7d2e-46be-821c-3e189d1c626c)
And i belive that is not a correct JS syntax.

I'll play around some with my limited knowledge to see if it's possible. 
Otherwise we might still need to do some parsing of the docstring on our end?
(Rr just skip the type anotations, but i do believe it would be real nice to have them)

ashish10alex commented:
@HampB , yh agree with all the points you raised. I'll be working on this further today to infer types the best way. I will put this PR on draft for now

HampB commented:
Sounds good, then i'll put this on hold aswell for now :-)

ashish10alex commented:
Hi @HampB , the PR is ready for testing again. I have updated the video demo to reflect the changes I have made. Changes since last time 

1. Use [comment-parser](https://www.npmjs.com/package/comment-parser) npm library to parse js-doc contents 
2. The doc string now shows function signature with arg and return types 
3. If a function does not have a doc string. It will show as its implementation as hover.

HampB commented on src/hoverProvider.ts:
```diff
@@ -23,6 +27,107 @@ const getMarkdownTableIdWtLink = (fullTableIdStruct:{database:string, schema:str
+        
+            const description = tag.description;
+
+            if(name === "This"){
```
Should this not be `if(tag.tag === "returns")` or similar?

HampB reviewed (commented):
Nicely done @ashish10alex, looking good. 

I've ran into an issue that i can see does not happen in your example video. 
For a function like this:

```javascript
  /**
    * This is a description of the function
    *
    * @param {Object} constants - this is constants
    * @param {Object} parameters - this is params
    * @returns {string} - this is the return
    */
  function myLocalFunction(constants, parameters) {
    return "myLocalFunction"
  };
```

I get the following hover, that seems broken:
![image](https://github.com/user-attachments/assets/8a43f836-4c5a-4905-a53c-bb0fe0a4f9d0)

I've added a comment on how i made it work for my example locally. But maybe i'm missing some scenario here?

ashish10alex commented:
Hi @HampB , thanks for catching that. Not sure why I had was not working in yours but the suggestion works and is the correct way of doing it. 

--

Could you please merge change from main to your branch before you merge this branch into yours ?
